Managers/MenuMan.cpp

```diff
@@ -0,0 +1,264 @@
+		}
+		m_TitleScreen = std::make_unique<TitleScreen>(m_GUIScreen.get());
+		m_MainMenu = std::make_unique<MainMenuGUI>(m_GUIScreen.get(), m_GUIInput.get());
+		m_ScenarioMenu = std::make_unique<ScenarioGUI>(m_GUIScreen.get(), m_GUIInput.get());
+
+		// TODO: MetaGameGUI doesn't seem to actually do anything with the Controller but removing conflicts with the second Create() method so that needs to be sorted out sometime in the year 3000.
+		m_MenuController = std::make_unique<Controller>(Controller::CIM_PLAYER);
+		g_MetaMan.GetGUI()->Create(m_MenuController.get());
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void MenuMan::Reinitialize() {
+		g_ConsoleMan.Destroy();
+		g_MetaMan.GetGUI()->Destroy();
+
+		m_ScenarioMenu.reset();
+		m_MainMenu.reset();
+		m_TitleScreen.reset();
+		m_MenuController.reset();
+
+		Initialize(false);
+		g_ConsoleMan.Initialize();
+
+		g_FrameMan.DestroyTempBackBuffers();
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void MenuMan::SetActiveMenu() {
+		ActiveMenu newActiveMenu = ActiveMenu::MenusDisabled;
+
+		switch (m_TitleScreen->GetTitleTransitionState()) {
+			case TitleScreen::TitleTransition::MainMenu:
+			case TitleScreen::TitleTransition::MainMenuToCredits:
+				newActiveMenu = ActiveMenu::MainMenuActive;
+				break;
+			case TitleScreen::TitleTransition::ScenarioMenu:
+				newActiveMenu = ActiveMenu::ScenarioMenuActive;
+				break;
+			case TitleScreen::TitleTransition::CampaignMenu:
+				newActiveMenu = ActiveMenu::CampaignMenuActive;
+				break;
+			default:
+				break;
+		}
+
+		if (newActiveMenu != m_ActiveMenu) {
+			m_ActiveMenu = newActiveMenu;
+			switch (m_ActiveMenu) {
+				case ActiveMenu::ScenarioMenuActive:
+					m_ScenarioMenu->SetEnabled(m_TitleScreen->GetPlanetPos(), m_TitleScreen->GetPlanetRadius());
+					break;
+				case ActiveMenu::CampaignMenuActive:
+					g_MetaMan.GetGUI()->SetPlanetInfo(m_TitleScreen->GetPlanetPos(), m_TitleScreen->GetPlanetRadius());
+					g_MetaMan.GetGUI()->SetEnabled();
+					break;
+				default:
+					break;
+			}
+		}
+	}
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	bool MenuMan::Update() {
+		if (g_FrameMan.ResolutionChanged()) { Reinitialize(); }
+
+		m_TitleScreen->Update();
+		SetActiveMenu();
+
+		bool quitResult = false;
+
+		switch (m_ActiveMenu) {
+			case ActiveMenu::MainMenuActive:
+				quitResult = UpdateMainMenu();
+				break;
+			case ActiveMenu::ScenarioMenuActive:
+				UpdateScenarioMenu();
+				break;
+			case ActiveMenu::CampaignMenuActive:
+				quitResult = UpdateCampaignMenu();
+				break;
+			default:
+				break;
+		}
+		if (!quitResult && m_TitleScreen->GetTitleTransitionState() != TitleScreen::TitleTransition::ScrollingFadeOutQuit) {
+			g_ConsoleMan.Update();
```

Comment:
I can actually pull out the updating and drawing out of here and throw it in `RunMenuLoop()` without any trouble.
I probably needed them here because of how things were set up earlier, don't remember.

Reply:
I'd prefer that, yeah. Makes more sense for the big overarching main game loop stuff to deal with consoleman, given consoleman is pretty separated from everything else going on.
That said, where is consoleman being updated when not in menus? Could it not be consolidated into somewhere in RunGameLoop or whatever, so only one bit of code has to update it?

Reply:
There are 2 main loops.
One for everything out-of-game (ie menus) and one for everything in-game.
They are completely separate and can't really be consolidated, at least not in a way that makes sense.
ConsoleMan being the little shit it is, is being updated in both of them.

Reply:
Oh yeah, I see. Should probably have refreshed my memory on the code there.
Only way I can see to consolidate it to one call is move the while loop out of the RunMenuLoop and RunGameLoop methods and into the `main` function, but that'd be way more work than this is worth lol.

Anyway, I'd say pull consoleman update out into RunMenuLoop and that'll deal with this issue just fine.

Reply:
The only way this could maybe work from within main is with fucky `goto`s.

The way things are set up is if there is no Activity at boot (ie not loading into Activity), then main() calls the menu loop.
When we start an Activity from the menus, we break from the menu loop and main() proceeds to run the game loop.

If there was an Activity at boot then the menu loop is skipped completely and we go straight into the game loop.
If from this point we want to go back to the menus we call the menu loop from the game loop, but we never break the game loop, cause that proceeds with executing main() and goes to cleanup and exit.

So if we called the main loop from the game loop, when we break out of it again it just proceeds with executing the game loop until we call the menu loop again, or until we quit.

Game loop basically becomes the main driving loop once you start any Activity.
It's a rather fucky setup that I don't particularly like, but can't do anything about.

Reply:
Goto deemed harmful lol.

Thanks for the explainer, that matches up with my understanding and adds some more helpful details. There are definitely ways to do it differently, and maybe better, but nothing obvious springs to mind, and I don't wanna spend time thinking about it. I agree that this setup is a bit fucky, but it's way better than things used to be, so I'm okay with it.